File: cplusplus/43MultiplyStrings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        if (num1.size() == 0 || num2.size() == 0)
            return "0";
        
        vector<int> res = vector<int>(num1.size() + num2.size(), 0);
        int offset = 1;
        int carrier = 0;
        for (int i = num1.size() - 1; i >= 0; --i) {
            int n1 = num1.at(i) - '0';
            for (int j = num2.size() - 1; j >= 0; --j) {
                int n2 = num2.at(j) - '0';
                int prod = n1 * n2;
                
                res[num1.size() - offset + 1 + j] += prod + carrier;
                carrier = res[num1.size() - offset + 1 + j] / 10;
                res[num1.size() - offset + 1 + j] %= 10;
            }
            
            int pos = num1.size() - offset;
            while (carrier > 0) {
                res[pos] += carrier;
                res[pos] %= 10;
                carrier /= 10;
                --pos;
            }

            ++offset;
        }

        string s = "";
        int i = 0;
        while (res[i] == 0 && i < res.size() - 1)
            ++i;
        
        for (; i < res.size(); ++i)
            s += '0' + res[i];
        
        return s;
    }
};

File: java/43MultiplyStrings.java
class Solution {
    public String multiply(String num1, String num2) {
        if (num1 == null || num1.length() == 0 ||
           num2 == null || num2.length() == 0)
            return "0";
        
        int[] product = new int[num1.length() + num2.length()];
        for (int i = num1.length() - 1; i >= 0; --i) {
            int carry = 0;
            for (int j = num2.length() - 1; j >= 0; --j) {
                char c1 = num1.charAt(i);
                char c2 = num2.charAt(j);
                product[i + j + 1] += (c1 - '0') *  (c2 - '0') + carry;
                carry = product[i + j + 1] / 10;
                product[i + j + 1] %= 10;
            }
            
            if (carry > 0)
                product[i] = carry; 
        }
        
        int i = 0;
        while (i < num1.length() + num2.length() && product[i] == 0)
            ++i;

        if (i == num1.length() + num2.length())
            return "0";
        String res = "" + (char)(product[i++] + '0');
        while (i < num1.length() + num2.length())
            res += (char)(product[i++] + '0');
        
        return res;
    }
}

// faster:

class Solution {
    public String multiply(String num1, String num2) {
        if (num1 == null || num1.length() == 0 ||
           num2 == null || num2.length() == 0)
            return "0";

        char[] n1 = num1.toCharArray();
        char[] n2 = num2.toCharArray();
        
        int[] product = new int[n1.length + n2.length];
        for (int i = n1.length - 1; i >= 0; --i) {
            int carry = 0;
            for (int j = n2.length - 1; j >= 0; --j) {
                char c1 = n1[i];
                char c2 = n2[j];
                product[i + j + 1] += (c1 - '0') *  (c2 - '0') + carry;
                carry = product[i + j + 1] / 10;
                product[i + j + 1] %= 10;
            }

            if (carry > 0)
                product[i] = carry;
        }

        int i = 0;
        while (i < n1.length + n2.length && product[i] == 0)
            ++i;

        if (i == n1.length + n2.length)
            return "0";
        String res = "" + (char)(product[i++] + '0');
        while (i < n1.length + n2.length)
            res += (char)(product[i++] + '0');

        return res;     
    }
}
